lib/ddtrace/profiling/ext/cpu.rb

```diff
@@ -19,6 +20,65 @@ def self.apply!
           # will provide a thread/clock ID for CPU timing.
           require 'ddtrace/profiling/ext/cthread'
           ::Thread.send(:prepend, Profiling::Ext::CThread)
+
+          # Applying hooks will allow any existing threads to update
+          # their thread/clock IDs so their CPU timings can be measured.
+          apply_hooks!
+        end
+
+        def self.apply_hooks!
+          # Threads that have already been created, will not have resolved
+          # a thread/clock ID. This is because these IDs can only be resolved
+          # from within the thread's execution context, which we do not control.
+          #
+          # We can work around this by applying trace hooks to capture each thread
+          # from within its own execution context, resolving the thread/clock IDs,
+          # then disabling the hook so it doesn't re-run needlessly.
+          Thread.list.each do |thread|
+            # Only attempt to hook into threads that are instrumented
+            next unless thread.is_a?(CThread)
+
+            # TracePoint supports filtering by thread for Ruby 2.7+
+            if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.7') && defined?(TracePoint)
+              thread[THREAD_HOOK] = TracePoint.new(:line) do |tp|
+                current_thread = Thread.current
+
+                begin
+                  current_thread.send(:update_native_ids)
+                rescue StandardError => e
+                  log "ERROR: Failed to update thread/clock IDs for Thread #{current_thread.object_id}.\nCause: #{e.message}\nLocation: #{e.backtrace.first}"
```

Comment:
To be veeery safe we should probably `rescue` any errors we get from trying to log as well, just in case.

Reply:
We could, but if we disable first before doing anything else, I'm less worried about this. It was more of an issue when disabling was happening in the `ensure`.

Reply:
Note that any unhandled exceptions inside the tracepoint block escape to the interrupted thread:

```ruby
background_thread = Thread.new do
  Thread.stop
  puts "Woken up!"
  :result
end

sleep 0.1 while background_thread.status == 'run'
background_thread.run

puts "Result is: #{background_thread.value}"

background_thread = Thread.new do
  Thread.stop
  puts "Woken up!"
  :result
end

TracePoint.new(:line) do
  raise 'Kaboom'
end.enable(target_thread: background_thread)

sleep 0.1 while background_thread.status == 'run'
background_thread.run

puts "Result is: #{background_thread.value}"
```

Result:

```
$ ruby tracepoint.rb
Woken up!
Result is: result
#<Thread:0x00007f8d2c874310 tracepoint.rb:13 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	1: from tracepoint.rb:14:in `block in <main>'
tracepoint.rb:20:in `block in <main>': Kaboom (RuntimeError)
Traceback (most recent call last):
	1: from tracepoint.rb:24:in `<main>'
tracepoint.rb:24:in `run': killed thread (ThreadError)
```